osu.Server.Spectator/Hubs/ServerMultiplayerRoom.cs

```diff
@@ -72,5 +89,112 @@ private MatchTypeImplementation createTypeImplementation(MatchType type)
+                catch
+                {
+                    // Any failures in the last countdown should not prevent future countdowns from running.
+                }
+
+                // Notify users that a new countdown has started.
+                // Note: The room must be re-retrieved rather than using our own instance to enforce single-thread access.
+                using (var roomUsage = await hubCallbacks.GetRoom(RoomID))
+                {
+                    if (roomUsage.Item == null)
+                        return;
+
+                    // The countdown could have been cancelled in a separate request before this task was able to run.
+                    if (stopSource.IsCancellationRequested)
+                        return;
+
+                    roomUsage.Item.Countdown = countdown;
+
+                    countdownStartTime = DateTimeOffset.Now;
+                    countdownDuration = countdown.TimeRemaining;
+
+                    await hubCallbacks.SendMatchEvent(roomUsage.Item, new CountdownChangedEvent { Countdown = countdown });
+                }
+
+                // Run the countdown.
+                try
+                {
+                    await Task.Delay(countdownDuration, cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Clients need to be notified of cancellations in the following code.
+                }
+
+                // Notify users that the countdown has finished (or cancelled) and run the continuation.
+                // Note: The room must be re-retrieved rather than using our own instance to enforce single-thread access.
+                using (var roomUsage = await hubCallbacks.GetRoom(RoomID))
+                {
+                    if (roomUsage.Item == null)
+                        return;
+
+                    roomUsage.Item.Countdown = null;
+
+                    await hubCallbacks.SendMatchEvent(roomUsage.Item, new CountdownChangedEvent { Countdown = null });
+
+                    using (cancellationSource)
+                    {
+                        if (stopSource.Token.IsCancellationRequested)
+                            return;
+                    }
+
+                    // The continuation could be run outside of the room lock, however it seems saner to run it within the same lock as the cancellation token usage.
+                    // Furthermore, providing a room-id instead of the room becomes cumbersome for usages, so this also provides a nicer API.
+                    await onComplete(roomUsage.Item);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stops the current countdown. Its continuation will not run.
+        /// </summary>
+        public void StopCountdown() => countdownStopSource?.Cancel();
+
+        /// <summary>
+        /// Forces the current countdown to finish and run its continuation as soon as possible.
```

Comment:
comment is still a big ambiguous (what does "finish" mean here? to me it means it's going to run the "start" call). How does this sound, if it matches the intention:

```suggestion
        /// Forces the current countdown to run its callback immediately, unless cancelled.
```

Reply:
It's not immediately. It's intended to read "as soon as possible". You can't assume the countdown is finished by the point this method returns.

Reply:
That's fine to change then. important part to me is the "unless cancelled" and not using the word "finish".

Reply:
Reworded xmldoc and matched with osu-side changes.